### PR TITLE
N°541 - Dashlets: Improve readability when to much labels (pie chart) or too long labels (bar chart)

### PR DESCRIPTION
**Pull request by accognet**
For bar chart: set chart size depending on label size
For pie chart : increase the size of the chart proportionnaly to number of labels if there are more than 50 labels (5 columns) in the legend

Before :
image
After :
image

### CODE REVIEW

File: application/displayblock.class.inc.php [hunk 1]
```diff
@@ -1009,14 +1009,18 @@ public function GetRenderContent(WebPage $oPage, array $aExtraParams, $sId)
...
 				foreach ($aRes as $iRow => $aRow)
 				{
 					$sValue = $aRow['grouped_by_1'];
 					$sHtmlValue = $oGroupByExp->MakeValueLabel($this->m_oFilter, $sValue, $sValue);
 					$aGroupBy[(int)$iRow] = (int) $aRow[$sFctVar];
 					$iTotalCount += $aRow['_itop_count_'];
 					$aValues[] = array('label' => html_entity_decode(strip_tags($sHtmlValue), ENT_QUOTES, 'UTF-8'), 'label_html' => $sHtmlValue, 'value' => (int) $aRow[$sFctVar]);
-					
+					if ($iMaxNbCharInLabel < strlen($sValue)) {
+						$iMaxNbCharInLabel = strlen($sValue);
```
Comment (Molkobain) [inline application/displayblock.class.inc.php]:
```suggestion
					if ($iMaxNbCharInLabel < strlen($sValue)) {
						$iMaxNbCharInLabel = strlen($sValue);
```

Rather use [mb_strlen](https://www.php.net/manual/fr/function.mb-strlen.php) to ensure it works as expected with non latin alphabets.

File: application/displayblock.class.inc.php [hunk 2]
```diff
@@ -1039,7 +1043,8 @@ public function GetRenderContent(WebPage $oPage, array $aExtraParams, $sId)
 				$sJson = json_encode($aValues);
 				$oPage->add_ready_script(
 <<<EOF
-
+var iSizeOfChar = (200 + 6 * $iMaxNbCharInLabel);
```
Comment (Molkobain) [inline application/displayblock.class.inc.php]: Add comment to explain the calculation

File: application/displayblock.class.inc.php [hunk 3]
```diff
@@ -1039,7 +1043,8 @@ public function GetRenderContent(WebPage $oPage, array $aExtraParams, $sId)
...
+$('#my_chart_$sId').height(iSizeOfChar + 'px');
```
Comment (Molkobain) [inline application/displayblock.class.inc.php]:
```suggestion
var iChartHeight = (200 + 6 * $iMaxNbCharsInLabel);
$('#my_chart_$sId').height(iChartHeight + 'px');
```

File: application/displayblock.class.inc.php [hunk 4]
```diff
@@ -1009,14 +1009,18 @@ public function GetRenderContent(WebPage $oPage, array $aExtraParams, $sId)
 				$iTotalCount = 0;
 				$aValues = array();
 				$aURLs = array();
+				$iMaxNbCharInLabel = 0;
```
Comment (Molkobain) [inline application/displayblock.class.inc.php]:
```suggestion
				$iMaxNbCharsInLabel = 0;
```

File: application/displayblock.class.inc.php (hunk 1 above)
Comment (Molkobain) [inline application/displayblock.class.inc.php]:
```suggestion
					if ($iMaxNbCharsInLabel < strlen($sValue)) {
						$iMaxNbCharsInLabel = strlen($sValue);
```

File: application/displayblock.class.inc.php [hunk 5]
```diff
@@ -1107,8 +1112,15 @@ public function GetRenderContent(WebPage $oPage, array $aExtraParams, $sId)
 				}
 				$sJSColumns = json_encode($aColumns);
 				$sJSNames = json_encode($aNames);
+				$iNbLineToAddForName = 0;
+				if( count($aNames)>50) {
+					//increase size of the graph in order to have a maximum of 5 columns of legend
+					$iNbLineToAddForName = ceil ( count($aNames) / 5) - 10;
```
Comment (Molkobain) [inline application/displayblock.class.inc.php]:
```suggestion
				$iNbLinesToAddForName = 0;
				if (count($aNames) > 50) {
					// Increase size of the graph in order to have a maximum of 5 columns of legend
					$iNbLinesToAddForName = ceil(count($aNames) / 5) - 10;
```

Comment (Molkobain) [inline application/displayblock.class.inc.php]: Add comment to explain the calculation

File: application/displayblock.class.inc.php [hunk 6]
```diff
@@ -1107,8 +1112,15 @@ public function GetRenderContent(WebPage $oPage, array $aExtraParams, $sId)
...
+				}
 				$oPage->add_ready_script(
 <<<EOF
+var iSizeOfChar = (200 + 20 * $iNbLineToAddForName);
+$('#my_chart_$sId').height(iSizeOfChar + 'px');
```
Comment (Molkobain) [inline application/displayblock.class.inc.php]:
```suggestion
var iChartHeight = (200 + 20 * $iNbLinesToAddForName);
$('#my_chart_$sId').height(iChartHeight + 'px');
```

**Review by Molkobain — CHANGES_REQUESTED** [review]

**Comment by Molkobain** [comment]
Accepted during technical review once changes have been made.
Anne-Catherine will add before/after screenshots to the description to ease discussion.

File: application/displayblock.class.inc.php [hunk 7]
```diff
@@ -1009,14 +1009,18 @@ public function GetRenderContent(WebPage $oPage, array $aExtraParams, $sId)
 				$iTotalCount = 0;
 				$aValues = array();
 				$aURLs = array();
+				$iMaxNbCharsInLabel = 0;
```
Comment (Molkobain) [inline application/displayblock.class.inc.php:1012]:
```suggestion
				$iMaxNbCharsInLabel = 0;
				
```

File: application/displayblock.class.inc.php [hunk 8]
```diff
@@ -1009,14 +1009,18 @@ public function GetRenderContent(WebPage $oPage, array $aExtraParams, $sId)
...
 				foreach ($aRes as $iRow => $aRow)
 				{
 					$sValue = $aRow['grouped_by_1'];
 					$sHtmlValue = $oGroupByExp->MakeValueLabel($this->m_oFilter, $sValue, $sValue);
 					$aGroupBy[(int)$iRow] = (int) $aRow[$sFctVar];
 					$iTotalCount += $aRow['_itop_count_'];
 					$aValues[] = array('label' => html_entity_decode(strip_tags($sHtmlValue), ENT_QUOTES, 'UTF-8'), 'label_html' => $sHtmlValue, 'value' => (int) $aRow[$sFctVar]);
-					
+					if ($iMaxNbCharsInLabel < mb_strlen($sValue)) {
```
Comment (Molkobain) [inline application/displayblock.class.inc.php:1022]:
```suggestion

					if ($iMaxNbCharsInLabel < mb_strlen($sValue)) {
```

File: application/displayblock.class.inc.php [hunk 9]
```diff
@@ -1039,7 +1043,9 @@ public function GetRenderContent(WebPage $oPage, array $aExtraParams, $sId)
 				$sJson = json_encode($aValues);
 				$oPage->add_ready_script(
 <<<EOF
-
+//calculate height of chart : 200px (height of the chart) + 6*iMaxNbCharsInLabel for the legend
```
Comment (Molkobain) [inline application/displayblock.class.inc.php]:
```suggestion
// Calculate height of chart : 200px (height of the chart) + 6*iMaxNbCharsInLabel for the legend
```

File: application/displayblock.class.inc.php [hunk 10]
```diff
@@ -1107,8 +1113,17 @@ public function GetRenderContent(WebPage $oPage, array $aExtraParams, $sId)
...
 				$sJSNames = json_encode($aNames);
+				$iNbLinesToAddForName = 0;
+				if (count($aNames) > 50) {
+					// Calculation of the number of legends line add to the height of the graph to have a maximum of 5 legend columns
+					// 10 corresponds to the number of lines already included in the chart height
+					$iNbLinesToAddForName = ceil(count($aNames) / 5) - 10;
+				}
 				$oPage->add_ready_script(
 <<<EOF
+//calculate height of graph : 200px (minimum height for the chart) + 20*iNbLinesToAddForName for the legend
```
Comment (Molkobain) [inline application/displayblock.class.inc.php]:
```suggestion
// Calculate height of graph : 200px (minimum height for the chart) + 20*iNbLinesToAddForName for the legend
```

File: application/displayblock.class.inc.php [hunk 11]
```diff
@@ -1039,7 +1043,9 @@ public function GetRenderContent(WebPage $oPage, array $aExtraParams, $sId)
 				$sJson = json_encode($aValues);
 				$oPage->add_ready_script(
 <<<EOF
-
+//calculate height of chart : 200px (height of the chart) + 6*iMaxNbCharsInLabel for the legend
+var iHeightOfChar = (200 + 6 * $iMaxNbCharsInLabel);
+$('#my_chart_$sId').height(iHeightOfChar + 'px');
```
Comment (Molkobain) [inline application/displayblock.class.inc.php]:
```suggestion
var iChartHeight = (200 + 6 * $iMaxNbCharsInLabel);
$('#my_chart_$sId').height(iChartHeight + 'px');
```

**Review by Molkobain — APPROVED** [review]

File: application/displayblock.class.inc.php [hunk 12]
```diff
@@ -1039,7 +1045,9 @@ public function GetRenderContent(WebPage $oPage, array $aExtraParams, $sId)
 				$sJson = json_encode($aValues);
 				$oPage->add_ready_script(
 <<<EOF
-
+// Calculate height of chart : 200px (height of the chart) + 6*iMaxNbCharsInLabel for the legend
+var iChartHeight = (200 + 6 * $iMaxNbCharsInLabel);
+$('#my_chart_$sId').height(iChartHeight + 'px');
```
Comment (piRGoif) [inline application/displayblock.class.inc.php]:
```suggestion
var iChartDefaultHeight = 200,
      iChartLegendHeight = 6 * $iMaxNbCharsInLabel,
      iChartTotalHeight = iChartDefaultHeight + iChartLegendHeight;
$('#my_chart_$sId').height(iChartTotalHeight+ 'px');
```

File: application/displayblock.class.inc.php [hunk 13]
```diff
@@ -1107,8 +1115,17 @@ public function GetRenderContent(WebPage $oPage, array $aExtraParams, $sId)
 				}
 				$sJSColumns = json_encode($aColumns);
 				$sJSNames = json_encode($aNames);
+				$iNbLinesToAddForName = 0;
+				if (count($aNames) > 50) {
+					// Calculation of the number of legends line add to the height of the graph to have a maximum of 5 legend columns
+					// 10 corresponds to the number of lines already included in the chart height
+					$iNbLinesToAddForName = ceil(count($aNames) / 5) - 10;
```
Comment (piRGoif) [inline application/displayblock.class.inc.php]:
```suggestion
					// Calculation of the number of legends line add to the height of the graph to have a maximum of 5 legend columns
					$iNbLinesIncludedInChartHeight = 10;
					$iNbLinesToAddForName = ceil(count($aNames) / 5) - $iNbLinesIncludedInChartHeight;
```

File: application/displayblock.class.inc.php [hunk 14]
```diff
@@ -1107,8 +1115,17 @@ public function GetRenderContent(WebPage $oPage, array $aExtraParams, $sId)
...
+				}
 				$oPage->add_ready_script(
 <<<EOF
+// Calculate height of graph : 200px (minimum height for the chart) + 20*iNbLinesToAddForName for the legend
+var iChartHeight = (200 + 20 * $iNbLinesToAddForName);
+$('#my_chart_$sId').height(iChartHeight + 'px');
```
Comment (piRGoif) [inline application/displayblock.class.inc.php]:
```suggestion
var iChartDefaultHeight = 200,
      iChartLegendHeight = 20 * $iNbLinesToAddForName,
      iChartTotalHeight = (200 + iChartLegendHeight);
$('#my_chart_$sId').height(iChartTotalHeight + 'px');
```

**Review by piRGoif — CHANGES_REQUESTED** [review]
As said during the review : inline comments can be lost easily after some modifications done to the code ! So it is a good practice to replace them with readable code when applicable.

Here we can introduce intermediate variables to explain the meaning of each block, so I would definitively prefer that : it will be far better in the long run, and even clearer for the developers.

**Comment by piRGoif** [comment]
Functional review : OK for 2.7
@accognet can you check my review before merging ?

**Pull request merged**